### PR TITLE
Add PNG exporting feature using cairo

### PR DESCRIPTION
**Pull request by mont5piques**
cairo is already installed (pygobject3); so no additionnal dependecy is required

### CODE REVIEW

File: tools/raw-log-converter.py [hunk 1]
```diff
@@ -1,4 +1,4 @@
-#!/bin/env python3
+#!/usr/bin/env python3
```
Comment (whot) [inline tools/raw-log-converter.py]: can you send this as separate patch please, this one is just a bug that needs to be fixed anyway.

File: tuhi/gui/drawing.py [hunk 2]
```diff
@@ -41,10 +43,12 @@ def __init__(self, json_data, zoom, *args, **kwargs):
         self.orientation = Config().orientation
         Config().connect('notify::orientation', self._on_orientation_changed)
         DATA_PATH.mkdir(parents=True, exist_ok=True)
+        PNG_DATA_PATH.mkdir(parents=True, exist_ok=True)
 
         self.json_data = json_data
         self._zoom = zoom
         self.process_svg()  # sets self.svg
+        self.process_png()  # sets self.png
```
Comment (whot) [inline tuhi/gui/drawing.py]: I don't think this one is needed. we don't use the png in the GUI directly, so we can really just generate those on-demand before export.

File: tuhi/gui/drawing.py [hunk 3]
```diff
@@ -53,6 +57,7 @@ def __init__(self, json_data, zoom, *args, **kwargs):
     def _on_orientation_changed(self, config, pspec):
         self.orientation = config.orientation
         self.process_svg()
+        self.process_png()
```
Comment (whot) [inline tuhi/gui/drawing.py]: same here

File: tuhi/gui/drawing.py [hunk 4]
```diff
@@ -115,21 +124,33 @@ def _on_download_button_clicked(self, button):
         # Translators: filter to show svg files only
         filter_svg.set_name(_('SVG files'))
         filter_svg.add_pattern('*.svg')
+        filter_png = Gtk.FileFilter()
+        # Translators: filter to show svg files only
```
Comment (whot) [inline tuhi/gui/drawing.py]: this would need updating

File: tuhi/gui/drawing.py [hunk 5]
```diff
@@ -115,21 +124,33 @@ def _on_download_button_clicked(self, button):
...
             file = dialog.get_filename()
-            shutil.copyfile(self.svg.filename, file)
-            # FIXME: error handling
+
+            if file.lower().endswith('.png'):
+                # regenerate the PNG based on the current rotation.
+                # where we used the orientation buttons, we haven't updated the
+                # file itself.
+                self.process_png()
+                shutil.copyfile(self.png.filename, file)
```
Comment (whot) [inline tuhi/gui/drawing.py]: I'd argue that after copying the file we should delete it from our own data path. we use svg primarily in the GUI and that's why we keep those around, but png is a purely export feature so we can delete it after saving it.

File: tuhi/gui/drawing.py [hunk 6]
```diff
@@ -18,13 +18,15 @@
...
 from tuhi.svg import JsonSvg
+from tuhi.png import JsonPng
 
 import gi
 gi.require_version("Gtk", "3.0")
 from gi.repository import GObject, Gtk, GdkPixbuf, Gdk  # NOQA
 
 
 DATA_PATH = Path(xdg.BaseDirectory.xdg_cache_home, 'tuhi', 'svg')
+PNG_DATA_PATH = Path(xdg.BaseDirectory.xdg_cache_home, 'tuhi', 'png')
```
Comment (whot) [inline tuhi/gui/drawing.py]: rename `DATA_PATH` to `SVG_DATA_PATH` for consistency please (could be done in a separate commit before this one).

File: tools/kete.py [hunk 7]
```diff
@@ -30,13 +30,15 @@
 
 try:
     from tuhi.svg import JsonSvg
+    from tuhi.png import JsonPng
```
Comment (whot) [inline tools/kete.py]: honestly, I wouldn't mind just shoving the png bits into `tuhi.svg`. Or possibly renaming it to `tuhi.image` or something.

File: tuhi/png.py [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+        self.orientation = orientation.lower()
+        self._convert()
+
+    def _convert(self):
+        js = self.json
+        dimensions = js['dimensions']
+        if dimensions == [0, 0]:
+            width, height = 100, 100
+        else:
+            # Original dimensions are too big for SVG Standard
```
Comment (whot) [inline tuhi/png.py]: SVG? :)

File: tuhi/png.py [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+            # so we normalize them
+            width, height = dimensions[0] // 100, dimensions[1] // 100
+
+        if self.orientation in ['portrait', 'reverse-portrait']:
+            size = (height, width)
+        else:
+            size = (width, height )
```
Comment (whot) [inline tuhi/png.py]: space before ` )`. btw, run `flake-8` and that should complain about this. Or install `flake-8` and run `ninja test` and that does the same.

File: tuhi/png.py [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+                ctx.set_line_width(stroke_width)
+
+                if i == 0:
+                    ctx.move_to(x, y)
+                else:
+                    ctx.line_to(x, y)
+
+            ctx.stroke()
+
+        surface.write_to_png(self.filename)
```
Comment (whot) [inline tuhi/png.py]: I can't think of a much better way right now (other than callbacks and whatnot) but it very much seems that we *should* be sharing the code with the `JsonSvg`. This is 90% the same and will eventually become the source of minor discrepancies as we forget to fix.

Maybe a parent class for the converter that calls the interface to add points/strokes which is then handled in the filetype-specific subclass?

**Review by whot — CHANGES_REQUESTED** [review]
Looks good, thanks. A few minor nitpicks, the only bigger thing I'd like us to think about is how we can share the code between the svg and the png converter - duplicating that will just lead to errors in the longer run.

File: tuhi/png.py (hunk 8 above)
Comment (mont5piques) [reply tuhi/png.py]: Copy paste for sure

**Comment by whot** [comment]
fwiw, you can just force-push to a branch and then ping the maintainer in the existing PR, no need to file a new PR for every iteration of the patch.

**Comment by whot** [comment]
oh, and the explicit ping is necessary because github doesn't (didn't) always notify for force-pushes. In this case I got the email though so the ping is more a "it's ready now" notification.

**Pull request closed without merging**